---
jupyter:
  jupytext:
    text_representation:
      extension: .Rmd
      format_name: rmarkdown
      format_version: '1.2'
      jupytext_version: 1.3.0
  kernelspec:
    display_name: Python [conda env:.conda-vanderburg_oropharyngeal_cancer]
    language: python
    name: conda-env-.conda-vanderburg_oropharyngeal_cancer-py
---

```{python}
import scanpy as sc
import pandas as pd
import operator as op
from pyscenic.utils import load_motifs
from pyscenic.plotting import plot_binarization, plot_rss
from pyscenic.transform import df2regulons
from pyscenic.binarization import binarize
from pyscenic.aucell import aucell
from pyscenic.export import export2loom, add_scenic_metadata
import seaborn as sns
import numpy as np
from cytoolz import compose
from pyscenic.rss import regulon_specificity_scores
from pyscenic.plotting import plot_binarization, plot_rss
from matplotlib import pyplot as plt
```

```{python tags=c("parameters")}
regulon_file = "../../SCENICprotocol/results/reg.csv"
input_adata = "../results.old/05_prepare_de/adata.h5ad"
output_adata = ""
n_cpus = 32
```

## Compute cellular enrichment with AUCell
This procedure follows the jupyter notebooks from `SCENICprotocol`

```{python}
adata = sc.read_h5ad(input_adata)
# undo log-transformation
expr_mat = np.expm1(adata.to_df())
```

```{python}
df_motifs = load_motifs(regulon_file)
```

```{python}
df_motifs
```

```{python}
def contains(*elems):
    def f(context):
        return any(elem in context for elem in elems)

    return f
```

```{python}
# adjust here if we want only a subset
db_names = [
    "hg38__refseq-r80__10kb_up_and_down_tss.mc9nr",
    "hg38__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr",
]
```

```{python}
df_motifs.columns = df_motifs.columns.droplevel(0)
```

```{python}
# For the creation of regulons we only keep the 10-species databases and the activating modules. We also remove the
# enriched motifs for the modules that were created using the method 'weight>50.0%' (because these modules are not part
# of the default settings of modules_from_adjacencies anymore.
df_motifs2 = df_motifs[
    np.fromiter(
        map(compose(op.not_, contains("weight>50.0%")), df_motifs.Context),
        dtype=np.bool,
    )
    & np.fromiter(map(contains(*db_names), df_motifs.Context), dtype=np.bool)
    & np.fromiter(map(contains("activating"), df_motifs.Context), dtype=np.bool)
]
```

```{python}
df_motifs2.shape
```

```{python}
# We build regulons only using enriched motifs with a NES of 3.0 or higher; we take only directly annotated TFs or TF annotated
# for an orthologous gene into account; and we only keep regulons with at least 10 genes.
regulons = list(
    filter(
        lambda r: len(r) >= 10,
        df2regulons(
            df_motifs2[
                (df_motifs2["NES"] >= 3.0)
                & (
                    (df_motifs2["Annotation"] == "gene is directly annotated")
                    | (
                        df_motifs2["Annotation"].str.startswith(
                            "gene is orthologous to"
                        )
                        & df_motifs2["Annotation"].str.endswith(
                            "which is directly annotated for motif"
                        )
                    )
                )
            ]
        ),
    )
)
```

```{python}
auc_mtx = aucell(expr_mat, regulons, num_workers=n_cpus)
```

```{python}
bin_mtx, thresholds = binarize(auc_mtx, num_workers=n_cpus)
```

```{python}
sns.clustermap(bin_mtx.iloc[:3000, :].T)
```

#### Add the data to the scanpy object

```{python}
add_scenic_metadata(adata, auc_mtx, regulons)
adata.obsm["X_aucell_bin"] = bin_mtx
adata.uns["scenic_bin_thresholds"] = thresholds
```

## UMAP using gene expression data:

```{python}
sc.pl.umap(adata, color=["patient", "cell_type_coarse", "cell_type"])
```

```{python}
adata_tmp = adata.copy()
```

```{python}
sc.pp.neighbors(adata_tmp, use_rep="X_aucell", n_neighbors=10)
sc.tl.umap(adata_tmp)
sc.tl.leiden(adata_tmp)
```

```{python}
sc.pl.umap(adata_tmp, color=["patient", "cell_type_coarse", "cell_type"], ncols=2)
```

```{python}
sc.pl.umap(
    adata_tmp, color=["leiden", "CD8A", "FOXP3", "KLRC1", "CD4"], ncols=2, cmap="magma"
)
```

## Save result

```{python}
adata.write_h5ad(output_adata, compression="lzf")
```
